# Find sum of all primes below 2,000,000

In [24]:
from datetime import datetime, timedelta
import sys
import math

In [31]:
def print_in_place(str):
    print(str,end="\r", flush=True)

In [46]:
import pdb

In [96]:
primes = []
composite = [False, False,]

def init_seive(n):
    for i in range(2, n):
        composite.append(True)

    for i in range(2, math.floor(math.sqrt(n))+1):
        if composite[i]:
            print_in_place("starting i= %s"% i)
            j=i**2
            while j < n:
                composite[j] = False
                j+=i
            print_in_place("finished i= %s" % i)


def find_primes(n):
    init_seive(n)
    i=0
    debug = 1000
    for i, is_prime in enumerate(composite):
        if is_prime:
            primes.append(i)
            i+=1
            if((i % debug)==0):
                print("prime %s found %s" % (len(primes), i), end="\r", flush=True)

            
start = datetime.now()
find_primes(2000000)
end = datetime.now()
print("\n",end - start)
print()


prime 148794 found 1998000
 0:00:02.045111



In [95]:
primes[-1:]

[999999937]

This approach was able to find all the primes up to 1 billion in under 10 mins on my laptop.
The last number in this sequence is: 999999937 :)